packages/ai-jsx/src/batteries/use-tools.tsx

```diff
@@ -183,6 +190,85 @@ export interface UseToolsProps {
  * ```
  *
  */
-export function UseTools(props: UseToolsProps) {
+export async function* UseTools(props: UseToolsProps, { render }: RenderContext) {
+  try {
+    const rendered = yield* render(<UseToolsFunctionCall {...props} />);
+    return rendered;
+  } catch (e: any) {
+    if (e.code === ErrorCode.ChatModelDoesntSupportFunctions) {
+      return <UseToolsPromptEngineered {...props} />;
+    }
+    throw e;
+  }
+}
+
+/**
+ * An implementation of <UseTools/> that uses {@link ChatCompletion}'s {@link FunctionDefinition} capability to call
+ * functions. The chat model in scope must support Function Calls for this component.
+ */
+export async function* UseToolsFunctionCall(props: UseToolsProps, { render }: RenderContext) {
+  const messages = [
+    <SystemMessage>You are a smart agent that may use functions to answer a user question.</SystemMessage>,
+  ];
+  if (props.fallback) {
+    messages.push(
+      <SystemMessage>Here's the fallback strategy/message if something failed: {props.fallback}</SystemMessage>
+    );
+  }
+  messages.push(<UserMessage>{props.query}</UserMessage>);
+
+  do {
+    const model_response = (
+      <ChatCompletion
+        functionDefinitions={Object.entries(props.tools).map(([toolName, tool]) => ({
+          name: toolName,
+          description: tool.description,
+          parameters: tool.parameters,
+        }))}
```

Comment:
Should we just map the type of `props.tools` precisely the same as `functionDefinitions`?

Reply:
Good callout. I changed functionDefinitions to match with props.tools